### PR TITLE
RUMM-3103 remove global sdk core

### PR DESCRIPTION
**Pull request by xgouchet**
### What does this PR do?

Remove all references to the globalSdkCore and general static fields of the Datadog class.

### CODE REVIEW

**Review by urseberry — APPROVED** [review]

File: instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/rum/ResourceTrackingTest.kt [hunk 1]
```diff
@@ -92,6 +93,7 @@ internal class ResourceTrackingTest {
     }
 
     @Test
+    @Ignore
```
Comment (0xnm) [inline instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/rum/ResourceTrackingTest.kt:96]: should this be reverted?

File: instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/rules/MockServerActivityTestRule.kt [hunk 2]
```diff
@@ -81,8 +80,10 @@ internal open class MockServerActivityTestRule<T : Activity>(
 
                     return if (request.path == CONNECTION_ISSUE_PATH) {
                         MockResponse().setSocketPolicy(SocketPolicy.NO_RESPONSE)
-                    } else {
+                    } else if (request.path.endsWith("nyan-cat.gif")) {
```
Comment (0xnm) [inline instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/rules/MockServerActivityTestRule.kt:83]: seems leftover from debugging?

File: instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/log/ConsentPendingGrantedLogsTest.kt [hunk 3]
```diff
@@ -33,7 +32,7 @@ internal class ConsentPendingGrantedLogsTest : LogsTest() {
     @Test
     fun verifyActivityLogs() {
         // update the tracking consent
-        Datadog.setTrackingConsent(TrackingConsent.GRANTED)
+        Datadog.getInstance()?.setTrackingConsent(TrackingConsent.GRANTED)
```
Comment (0xnm) [inline instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/log/ConsentPendingGrantedLogsTest.kt:35]: I'm wondering if we can also have some non-null variant of getting the instance, because for example in nightly-tests we have to use `!!`, which is not very nice.

To me it is the same situation as getting the `Activity` in Android, where `getActivity` and `requireActivity` exist (which looks to me as a hacky workaround).

File: instrumented/nightly-tests/src/androidTest/kotlin/com/datadog/android/nightly/logs/LoggerE2ETests.kt [hunk 4]
```diff
@@ -62,14 +62,12 @@ class LoggerE2ETests {
...
      */
     @Before
     fun setUp() {
-        initializeSdk(
-            InstrumentationRegistry.getInstrumentation().targetContext,
-            forgeSeed = forge.seed
-        )
-        logger = Logger.Builder(Datadog.globalSdkCore)
+        initializeSdk(InstrumentationRegistry.getInstrumentation().targetContext)
+        logger = Logger.Builder(Datadog.getInstance()!!)
```
Comment (0xnm) [inline instrumented/nightly-tests/src/androidTest/kotlin/com/datadog/android/nightly/logs/LoggerE2ETests.kt:70]: Passing the seed it missing and we can use `sdkCore` return value.

```suggestion
        val sdkCore = initializeSdk(InstrumentationRegistry.getInstrumentation().targetContext, forgeSeed = forge.seed)
        checkNotNull(sdkCore)
        logger = Logger.Builder(sdkCore)
```

File: library/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/AndroidTracer.kt [hunk 5]
```diff
@@ -226,6 +227,7 @@ class AndroidTracer internal constructor(
             val propagationStyles = tracingHeaderTypes.joinToString(",")
             properties.setProperty(Config.PROPAGATION_STYLE_EXTRACT, propagationStyles)
             properties.setProperty(Config.PROPAGATION_STYLE_INJECT, propagationStyles)
+            properties.setProperty(Config.PRIORITY_SAMPLING, "false")
```
Comment (0xnm) [inline library/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/AndroidTracer.kt:230]: why this change is needed?

File: dd-sdk-android/src/main/kotlin/com/datadog/android/Datadog.kt [hunk 6]
```diff
@@ -149,36 +121,33 @@ object Datadog {
...
+            instance?.stop()
+        }
     }
 
+    // endregion
+
+    // region Global methods
+
     // Stop all Datadog work (for test purposes).
     internal fun stop() {
```
Comment (0xnm) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/Datadog.kt:149]: seems like this method can be removed, given that we can call `stopInstance` directly from public API.

File: instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/rules/MockServerActivityTestRule.kt [hunk 7]
```diff
@@ -141,13 +145,13 @@ internal open class MockServerActivityTestRule<T : Activity>(
     // region MockServerRule
 
     fun getRequests(): List<HandledRequest> {
-        Log.i(TAG, "Caught ${requests.size} requests")
+//        Log.i(TAG, "Caught ${requests.size} requests")
```
Comment (0xnm) [inline instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/rules/MockServerActivityTestRule.kt:148]: should be reverted here and below it seems

**Review by 0xnm — COMMENTED** [review]
nice work! I added few comments.

File: dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/data/upload/UploadWorker.kt [hunk 8]
```diff
@@ -10,21 +10,18 @@ import android.content.Context
...
-import com.datadog.android.v2.core.DatadogCore
+import com.datadog.android.v2.core.InternalSdkCore
 import com.datadog.android.v2.core.internal.net.DataUploader
 import java.util.LinkedList
 import java.util.Queue
 import java.util.concurrent.CountDownLatch
 import java.util.concurrent.TimeUnit
 
-@Suppress("DEPRECATION") // TODO RUMM-3103 remove deprecated references
 internal class UploadWorker(
```
[the remark this answers is not in the document]
Comment (xgouchet) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/data/upload/UploadWorker.kt:23]: Oh indeed that's an issue :/

File: dd-sdk-android/src/main/kotlin/com/datadog/android/error/internal/DatadogExceptionHandler.kt [hunk 9]
```diff
@@ -73,9 +71,8 @@ internal class DatadogExceptionHandler(
         // TODO RUMM-0000 If DatadogExceptionHandler goes into dedicated module (module of 1 class
         //  only?), we have to wait for the write in some other way
         // give some time to the persistence executor service to finish its tasks
-        val coreFeature = (Datadog.globalSdkCore as? DatadogCore)?.coreFeature
-        if (coreFeature != null) {
-            val idled = (coreFeature.persistenceExecutorService as? ThreadPoolExecutor)
+        if (sdkCore is InternalSdkCore) {
```
[the remark this answers is not in the document]
Comment (0xnm) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/error/internal/DatadogExceptionHandler.kt:74]: Let's keep it like that then, I'm not quite a fan of pulling this internal thing to the public `SdkCore`.

Regarding throwing an exception: my assumption is that if it is thrown, it will always be caught by our integration tests before leaking to the release, because it is us only who are creating the cores.

[the remark this answers is not in the document]
Comment (0xnm) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/error/internal/DatadogExceptionHandler.kt:74]: My concern is just to avoid places in the SDK where something gets unnoticed if it is not as what we expect, this can lead to the hard debugging, especially if it happens on the client side.

File: instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/log/ConsentPendingGrantedLogsTest.kt (hunk 3 above)
Comment (0xnm) [reply instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/log/ConsentPendingGrantedLogsTest.kt:35]: Yes, I'm not a fan of such API either. I'm just curious how we can reduce the number of nullable states in our codebase.

File: dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/data/upload/UploadWorker.kt [hunk 10]
```diff
@@ -33,38 +32,34 @@ internal class UploadWorker(
...
     override fun doWork(): Result {
-        if (!Datadog.isInitialized()) {
+        // the idea behind upload is the following:
+        // 1. we shuffle features list to randomize initial upload task sequence. It is done to
+        // avoid the possible bottleneck when some feature has big batches which are uploaded
+        // slowly, so that next time other features don't wait and have a chance to go before.
+        // 2. we introduce FIFO queue also to avoid the bottleneck: if some feature batch cannot
+        // be uploaded we put retry task to the end of queue, so that batches of other features
+        // have a chance to go.
+        val sdkCore = Datadog.getInstance() as? InternalSdkCore
```
Comment (0xnm) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/data/upload/UploadWorker.kt:42]: When we create `uploadWorkRequest` we can add the name of the instance it is created for and then read it here and get the instance needed.

**Review by 0xnm — APPROVED** [review]
lgtm! I've added a comment regarding `UploadWorker`, so that it is not associated with default instance only.
